pkg/pmem-csi-driver/nodeserver.go

```
@@ -88,57 +101,92 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	val, ok := req.GetVolumeContext()[volumeContextEphemeral]
+	if ok {
+		b, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
 		}
+		ephemeral = b
+	} else {
+		// If met all below conditions we still treat as a ephemeral volume request:
+		// 1) No Volume found with given volume id
+		// 2) No provisioner info found in VolumeContext "storage.kubernetes.io/csiProvisionerIdentity"
+		// 3) No StagingPath in the request
+		device, _ := ns.cs.dm.GetDevice(req.VolumeId)
+		_, ok := req.GetVolumeContext()[volumeProvisionerIdentity]
+		ephemeral = device == nil && !ok && len(srcPath) == 0
 	}
 
-	readOnly := req.GetReadonly()
-	attrib := req.GetVolumeContext()
-	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	var mountOptions []string
 
-	klog.V(3).Infof("NodePublishVolume: id:%s targetpath:%v stagingtargetpath:%v readonly:%v attributes:%v mountflags:%v",
-		req.GetVolumeId(), targetPath, stagingtargetPath, readOnly, attrib, mountFlags)
+	if ephemeral {
+		val, ok := req.GetVolumeContext()["size"]
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "size not found in volume context")
+		}
+
+		size, err := resource.ParseQuantity(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Failed to parse given size '%s': %s", val, err.Error()))
+		}
+
+		// Create new device
+		params := req.GetVolumeContext()
+		params[pmemParameterKeyPersistencyModel] = string(pmemPersistencyModelEphemeral)
+		createReq := &csi.CreateVolumeRequest{
+			VolumeCapabilities: []*csi.VolumeCapability{req.VolumeCapability},
+			Name:               req.VolumeId,
+			CapacityRange:      &csi.CapacityRange{RequiredBytes: size.Value()},
+			Parameters:         params,
+		}
 
-	options := []string{"bind"}
-	if readOnly {
-		options = append(options, "ro")
+		createResp, err := ns.cs.CreateVolume(ctx, createReq)
```

We need full validation of the volume parameters. Unknown (or miss-spelled) parameters must trigger an error. Otherwise user errors may silently lead to a different than intended outcome.

As a testcase, I tried with `foo: bar` added to `pmem-app-ephemeral.yaml` and that was silently ignored.

---

And we also need an automated test case for this.

---

Beware that we enable pod info on mount and thus have to ignore *all* `csi.storage.k8s.io/*` entries. Unknown fields may get added under that namespace in the future.

---

Done, Now we validate the volume context and raise an error if occurred any unexpected keys. Testcase is pending.